### PR TITLE
[feat] KValue base class

### PR DESCRIPTION
**Pull request by S1eGa**

### CODE REVIEW

File: include/klee/Module/KValue.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#define KVALUE_H
+
+namespace llvm {
+class Value;
+class StringRef;
+} // namespace llvm
+
+namespace klee {
+
+struct KValue {
```
Comment (misonijnik) [inline include/klee/Module/KValue.h:11]: Make KValue an abstract class.

File: lib/Expr/SymbolicSource.cpp [hunk 2]
```diff
@@ -102,8 +103,11 @@ int SymbolicSizeConstantAddressSource::internalCompare(
     return size < ub.size ? -1 : 1;
   }
   if (allocSite != ub.allocSite) {
-    return allocSite->getGlobalIndex() < ub.allocSite->getGlobalIndex() ? -1
-                                                                        : 1;
+    if (allocSite->getKind() == ub.allocSite->getKind()) {
```
Comment (misonijnik) [inline lib/Expr/SymbolicSource.cpp:106]: KVlaue comparisons must be made within this class.

File: lib/Module/KValue.cpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
+#include "klee/Support/CompilerWarning.h"
+#include <klee/Module/KValue.h>
```
Comment (misonijnik) [inline lib/Module/KValue.cpp:2]: < ... > -> " ... "

File: include/klee/Module/KModule.h [hunk 4]
```diff
@@ -84,6 +89,10 @@ struct KBlock {
 
   /// Block number in function
   [[nodiscard]] uintptr_t getId() const;
+
+  static bool classof(const KValue *rhs) {
+    return rhs->getKind() == Kind::BLOCK ? classof(cast<KBlock>(rhs)) : false;
```
Comment (misonijnik) [inline include/klee/Module/KModule.h:94]: Ternary operator -> &&

**Review by misonijnik — CHANGES_REQUESTED** [review]

File: lib/Solver/BitwuzlaSolver.cpp [hunk 5]
```diff
@@ -771,7 +771,7 @@ class BitwuzlaIncNativeSolver {
   Bitwuzla &getOrInit();
 
   bool isConsistent() const {
-    klee_warning("Empty isConsistent() check");
+    // klee_warning("Empty isConsistent() check");
```
Comment (ocelaiwo) [inline lib/Solver/BitwuzlaSolver.cpp:774]: I think we should attach a comment to this function that will describe what kind of check is to be expected here in the future (or remove this function altogether).

File: lib/Expr/SourceBuilder.cpp [hunk 6]
```diff
@@ -21,7 +22,7 @@ SourceBuilder::uninitialized(unsigned version, const KInstruction *allocSite) {
 }
 
 ref<SymbolicSource> SourceBuilder::symbolicSizeConstantAddress(
-    unsigned version, const KInstruction *allocSite, ref<Expr> size) {
+    unsigned version, const KValue *allocSite, ref<Expr> size) {
```
Comment (ocelaiwo) [inline lib/Expr/SourceBuilder.cpp:25]: We need some way to restrict `allocSite` to those types of `KValue` that make sense as alloc sites. At least make a comment somewhere describing accepted derivatives or make a set of constructors accepting those derivatives only.

**Review by ocelaiwo — CHANGES_REQUESTED** [review]

**Review by ocelaiwo — APPROVED** [review]

**Review by misonijnik — APPROVED** [review]

**Pull request merged**